Pie: Adding outer border, text position options

Billiam opened the pull request:
## :bookmark_tabs: Summary

This adds two small features to pie charts for better readability/reducing clutter.

First, it makes the distance of the label from the center of the pie chart user configurable, as a decimal from 0 to 1.

Second, it adds a circular path behind the pie slices, so that an overall border can be configured separately.

## :straight_ruler: Design Decisions

My goal was to allow pie charts to be a bit more readable and less cluttered. Even with very few options, there's very little room for labels that are still readable. 

Moving the labels further toward the outside of the pie (where the wedge is larger) gives them more room if needed.

Adding a separate style-able outer border allows the pie slice borders to be hidden, giving more room for text, while still allowing a circular border.

### Before

![Screenshot_20230224_181355](https://user-images.githubusercontent.com/242008/221326959-6f43220b-cb6f-4419-a9b8-71655d80bd21.png)

### After

(text position set to 80%, outside border width set to 4px, with pie slice borders styled off, bold text)

This isn't ideal because forest theme is very low contrast, and pie uses transparent slices by default.

![Screenshot_20230224_184219](https://user-images.githubusercontent.com/242008/221326975-b2b416b9-7384-47a2-8f34-b11ff6034206.png)

(with higher contrast color scheme)

![Screenshot_20230224_184133](https://user-images.githubusercontent.com/242008/221327022-b36a57aa-c3fe-410a-8d4b-ba2c211201ed.png)

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [x] :notebook: have added documentation (if appropriate)
- [x] :bookmark: targeted `develop` branch

sidharthv96 reviewed (changes requested):
Thanks @Billiam, looks good. 
Hiding the inner borders should be an option too. 
Need some more tests and documentation. 
You can also edit the `demos/pie.html` with some examples.
`packages/mermaid/src/docs/syntax/pie.md` needs some info about the new configs too.

Billiam commented on cypress/integration/rendering/pie.spec.js:
```diff
@@ -75,4 +75,20 @@ describe('Pie Chart', () => {
       expect(svg).to.not.have.attr('style');
     });
   });
+
+  it('should render a pie diagram with given outside stroke width', () => {
+    renderGraph(
+      `
+    pie title Sports in Sweden
+       "Bandy" : 40
+       "Ice-Hockey" : 80
+       "Football" : 90
+      `,
+      { pie: { outerBorderWidth: 5 } }
+    );
+    cy.get('.pieOuterCircle').should((circle) => {
+      const strokeWidth = parseFloat(circle.attr('stroke-width'));
+      expect(strokeWidth).to.eq(5);
+    });
+  });
```
I'd like to add one but, the only value changed by the text-position parameter are the label coordinates.

Any suggestions on what a test for this might look like, that doesn't duplicate the rendering code?

The X/Y positions is given by the arc generator, based on the user parameter and the radius
The radius is based on the width/height/margin

sidharthv96 replied:
You can add an `imgSnapshotTest`, which verifies the visual output.

Billiam replied:
Added

sidharthv96 replied:
Cool!

Billiam commented on packages/mermaid/src/config.type.ts:
```diff
@@ -222,7 +222,10 @@ export interface MindmapDiagramConfig extends BaseDiagramConfig {
 }
 
-export type PieDiagramConfig = BaseDiagramConfig;
+export interface PieDiagramConfig extends BaseDiagramConfig {
+  outerBorderWidth?: number;
+  textPosition?: number;
+}
```
Here, or in defaultConfig? I don't see any existing docs in this file currently

sidharthv96 commented:
> Hiding the inner borders should be an option too.

Is there a config for this like `outerBorderWidth`?
Maybe `innerBorderWidth`?

Billiam commented:
> > Hiding the inner borders should be an option too.
> 
> Is there a config for this like `outerBorderWidth`? Maybe `innerBorderWidth`?

I didn't add one, since there's already the `pieStrokeWidth` theme option...
Perhaps both textPosition and outerBorderWidth should also be theme variables?

sidharthv96 commented:
> I didn't add one, since there's already the pieStrokeWidth theme option...

I didn't know that existed. 😅 
In that case, outerBorderWidth might be a better fit in the theme option. 

textPosition can still be a config. And maybe we should set the default text position as 75% too.

Billiam commented:
> we should set the default text position as 75% too

Will that break existing pie snapshot tests? How are those invalidated/rebuilt?

Billiam commented:
> In that case, outerBorderWidth might be a better fit in the theme option.

Is it alright that `pieOuterStrokeWidth`  must be an integer (or decimal), while `pieStrokeWidth` (and other theme values) use CSS strings (like `'2px'`)?

sidharthv96 commented:
> > we should set the default text position as 75% too
> 
> Will that break existing pie snapshot tests? How are those invalidated/rebuilt?

That's done using Applitools platform. I'll take care of approving those changes. 

---

> Is it alright that pieOuterStrokeWidth must be an integer (or decimal), while pieStrokeWidth (and other theme values) use CSS strings (like '2px')?


I think we should be consistent (atleast for new options). Is there any technical difficulty in there?

Billiam commented:
> Is there any technical difficulty in there?

There is! The value sets both the border width, and expands the circle by half that width as well, so it needs to be dividable.

https://github.com/mermaid-js/mermaid/pull/4145/files#diff-f404c216a7c997119e7eef9a92a32fb29f1d36502d4ccfffeec2794a5e93e26cR125-R126

Otherwise, half the border width appears underneath the pie slices (which are transparent by default), and it doesn't look great.

sidharthv96 commented:
You can use the `parseFontSize` to get the number value.

Billiam commented:
> You can use the parseFontSize function in utils to get the number value.

Ah perfect. I've moved it to a theme variable and renamed it to `pieOuterStrokeWidth` to be more consistent with the other pie theme vars, and added it (and the other existing theme variables) to the theming document.

Billiam commented:
@sidharthv96  Are there any other changes needed for this PR?

sidharthv96 reviewed (approved):
Nope, LVGTM!
Thanks for the comprehensive docs update too!

Billiam commented:
Sweet, thanks very much!
